Restructuring ModelVersion and LocalModel

tidylobster opened the pull request:
* Introduced breaking changes to the API: class names, method names, method parameters, etc.
* Introduced unified request error handling for checking/raising client-side/server-side errors.
* Removed all code for parsing yaml definitions. It only makes sense to use it within CLI, so it will reside there.
* Removed UploadResponse object. Now LocalModel.upload returns ModelVersion instance directly.
* Removed Metricable class as it's no longer makes sense to carry arround a whole class for just two methods for a signle usage. 
* Removed Cluster.connect method in favor of pure class constructor. 
* Took out uploading metrics from LocalModel and transferred functionality to ModelVersion. Now there are two new methods on a ModelVersion instance: `.as_metric(threshold=10, comparator: ThresholdCmpOp.LESS_EQ)` and `.assign_metrics([metric])`. Instead of manually handling MetricSpecConfig, MetricSpec user can work with this:

    ```python
    mv1 = ModelVersion.find_by_id(cluster, 1)
    mv2 = ModelVersion.find_by_id(cluster, 2)
    mv2.assign_metrics([mv1.as_metric(10, ThresholdCmpOp.LESS_EQ)])
    ```
* Added functionality for uploading training data. Current implementation of `wait` is a subject of change in the future releases.

    ```python
    modelversion = ModelVersion.find(cluster, "my-model", 1)
    modelversion.training_data = "s3://bucket/path/to/training-data.csv"
    data_upload_response = modelversion.upload_training_data()
    data_upload_response.wait(retry=3, sleep=30)
    ```
* Improved/refactored tests regarding cluster entities (Application, ModelVersion, etc). 
* Added/updated class-level documentation with code-snippets for important Hydrosphere entities. 
* Fixed various misspellings and minor bugs.

Valenzione commented on hydrosdk/modelversion.py:
```diff
@@ -535,11 +593,80 @@ class NotFound(Exception):
 
 
-class UploadResponse:
+class ModelUploadResponse:
```
Personally, I am in favor of this renaming. Despite extending a name by a few characters it prevents any possible misunderstanding.

trthhrtz replied:
@Valenzione it creates ambiguity. At the very least there should be a hint in the docstring. Otherwise imho the name should be associated either with localModel or modelVersion as we don't have a model entity

Valenzione replied:
Ah, yeah, I missed that point, I got focused on `Why renaming?`. I agree with the ambiguity mentioned.

tidylobster replied:
I've introduced a new DataUploadResponse class, which means having UploadResponse might be confusing. ModelUploadResponse makes it more verbose

trthhrtz replied:
@tidylobster it's just that we don't have a Model entity and using it anywhere makes an ambiguity...

tidylobster commented on hydrosdk/modelversion.py:
```diff
@@ -535,11 +593,80 @@ class NotFound(Exception):
     class BadRequest(Exception):
         pass
 
-    class BadResponse(Exception):
```
BadResponse can appear not only for LocalModel/ModelVersion but for DataUpload as well. In my mind those are 5xx requests

tidylobster replied:
BadRequest (in the ModelVersion) on the other hand might signal that there's something wrong with the model definition, so server will return 4xx request.

tidylobster commented on hydrosdk/modelversion.py:
```diff
@@ -126,10 +176,10 @@ class LocalModel(Metricable):
 
     @staticmethod
-    def create(path, name, runtime, contract=None, install_command=None):
+    def create(name, runtime, payload, contract=None, install_command=None):
```
Yes, I've discussed this with @KineticCookie, this method should only be used for internal usage. E.g. when we just want to create an instance of LocalModel without reading serving.yaml, but specifying all fields ourselves. Need to create an issue for this one.

tidylobster commented on hydrosdk/modelversion.py:
```diff
@@ -489,6 +540,11 @@ def list_modelversions_by_model_name(cluster: Cluster, model_name: str) -> list:
 
         return sorted_by_version
+    
+    def upload_training_data(self) -> 'DataUploadResponse':
```
Data upload operates only on the modelversion, it doesn't make sense to utilize in LocalModel class

trthhrtz replied:
Okay, but why you did not do this:

```
        if self.training_data is not None:
            return upload_training_data(self.cluster, self.id, self.training_data)
        raise ValueError('Training data is not specified')
```
:)

tidylobster replied:
Do this where?

trthhrtz replied:
Ah, ok. I see your point. 

Moving forward, - 

1) I thought you said that both modelversion and localmodel will have these methods?

2) As a developer for me it's hard to understand the difference between 2 upload_training_data methods. When which should be used and which of them is internal?

tidylobster commented on hydrosdk/modelversion.py:
```diff
@@ -20,6 +21,51 @@
+    :param path: path to a local file
+    :param chunk_size: chunk size to use for streaming
+    """
+    def read_in_chunks(filename, chunk_size):
```
Well, I thought, since we are utilizing this function only within this method, it would be okay to have it as a subfunction

Valenzione commented on hydrosdk/exceptions.py:
```diff
@@ -1,3 +1,21 @@
+    pass
+
+
+class BadResponse(BaseException):
```
Python [documentation](https://docs.python.org/3/library/exceptions.html) states that 
>The built-in exception classes can be subclassed to define new exceptions; programmers are encouraged to derive new exceptions from the `Exception` class or one of its subclasses, and **not from** `BaseException`. More information on defining exceptions is available in the Python Tutorial under User-defined Exceptions.

I suggest following this guideline

tidylobster commented:
What has been done: 
* Introduced breaking changes to the API, hense class names, method names, method parameters, etc.
* Introduced unified request error handling for checking/raising client-side/server-side errors.
* Removed all code for parsing yaml definitions. It only makes sense to use it within CLI, so it will reside there.
* Removed UploadResponse object. Now LocalModel.upload returns ModelVersion instance directly.
* Removed Metricable class as it's no longer makes sense to carry arround a whole class for just two methods for a signle usage. 
* Removed Cluster.connect method in favor of pure class constructor. 
* Took out uploading metrics from LocalModel and transferred functionality to ModelVersion. Now there are two new methods on a ModelVersion instance: `.as_metric(threshold=10, comparator: ThresholdCmpOp.LESS_EQ)` and `.assign_metrics([metric])`. Instead of manually handling MetricSpecConfig, MetricSpec user can work with this:

    ```python
    mv1 = ModelVersion.find_by_id(cluster, 1)
    mv2 = ModelVersion.find_by_id(cluster, 2)
    mv2.assign_metrics([mv1.as_metric(10, ThresholdCmpOp.LESS_EQ)])
    ```
* Added functionality for uploading training data. Current implementation of `wait` is a subject of change in the future releases.

    ```python
    modelversion = ModelVersion.find(cluster, "my-model", 1)
    modelversion.training_data = "s3://bucket/path/to/training-data.csv"
    data_upload_response = modelversion.upload_training_data()
    data_upload_response.wait(retry=3, sleep=30)
    ```
* Improved/refactored tests regarding cluster entities (Application, ModelVersion, etc). 
* Added/updated class-level documentation with code-snippets for important Hydrosphere entities. 
* Fixed various misspellings and minor bugs.

tidylobster commented:
What has been done: 
* Introduced breaking changes to the API, hense class names, method names, method parameters, etc.
* Introduced unified request error handling for checking/raising client-side/server-side errors.
* Removed all code for parsing yaml definitions. It only makes sense to use it within CLI, so it will reside there.
* Removed UploadResponse object. Now LocalModel.upload returns ModelVersion instance directly.
* Removed Metricable class as it's no longer makes sense to carry arround a whole class for just two methods for a signle usage. 
* Removed Cluster.connect method in favor of pure class constructor. 
* Took out uploading metrics from LocalModel and transferred functionality to ModelVersion. Now there are two new methods on a ModelVersion instance: `.as_metric(threshold=10, comparator: ThresholdCmpOp.LESS_EQ)` and `.assign_metrics([metric])`. Instead of manually handling MetricSpecConfig, MetricSpec user can work with this:

    ```python
    mv1 = ModelVersion.find_by_id(cluster, 1)
    mv2 = ModelVersion.find_by_id(cluster, 2)
    mv2.assign_metrics([mv1.as_metric(10, ThresholdCmpOp.LESS_EQ)])
    ```
* Added functionality for uploading training data. Current implementation of `wait` is a subject of change in the future releases.

    ```python
    modelversion = ModelVersion.find(cluster, "my-model", 1)
    modelversion.training_data = "s3://bucket/path/to/training-data.csv"
    data_upload_response = modelversion.upload_training_data()
    data_upload_response.wait(retry=3, sleep=30)
    ```
* Improved/refactored tests regarding cluster entities (Application, ModelVersion, etc). 
* Added/updated class-level documentation with code-snippets for important Hydrosphere entities. 
* Fixed various misspellings and minor bugs.

Valenzione commented on tests/common_fixtures.py:
```diff
@@ -0,0 +1,185 @@
+
+@pytest.fixture(scope="session")
+def runtime():
+    return DockerImage("hydrosphere/serving-runtime-python-3.6", "2.1.0", None)
```
Why `2.1.0` and not `2.3.2`?

trthhrtz commented on hydrosdk/utils.py:
```diff
@@ -1,18 +1,56 @@
-import grpc
+from typing import Generator
 
-ALLOWED_TIMEOUT_gRPC_SECONDS = 5
```
why deleting default timeout?

Valenzione replied:
Because @tidylobster have moved timeout to the `Cluster` constructor

trthhrtz commented on tests/common_fixtures.py:
```diff
@@ -0,0 +1,185 @@
+from hydrosdk.image import DockerImage
+from hydrosdk.contract import ModelContract, SignatureBuilder
+from hydrosdk.modelversion import LocalModel
+from tests.config import *
```
> Wildcard imports (from <module> import *) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools. 
https://www.python.org/dev/peps/pep-0008/#imports

trthhrtz commented on tests/test_servable.py:
```diff
@@ -1,89 +1,81 @@
+    mv: ModelVersion = local_model.upload(cluster)
+    mv.lock_till_released()
+    sv: Servable = Servable.create(cluster, mv.name, mv.version)
+    assert sv.name in [servable.name for servable in Servable.list_all(cluster)]
```
error messages in case of assert is wrong would be helpful

Valenzione replied:
@trthhrtz Would you mind suggesting changes while @tidylobster is on vacation?

trthhrtz commented:
Great work 👍 . 


p.s. A bit difficult to code-review this large PR's 😄 

@KineticCookie  what do you think about
> Removed all code for parsing yaml definitions. It only makes sense to use it within CLI, so it will reside there.
?

Valenzione commented on hydrosdk/application.py:
```diff
@@ -22,135 +22,177 @@ class ApplicationStatus(Enum):
+        :return: application object
+        """
+        app_id = application_json.get("id")
+        app_name = application_json.get("name")
```
I think there are no such cases, wdyt @KineticCookie ?

trthhrtz replied:
@Valenzione If there are none I see it to be ambiguous sometimes to use `.get()` and sometimes to access dict's elements directly

Valenzione commented on hydrosdk/cluster.py:
```diff
@@ -68,22 +59,19 @@ def __init__(self, http_address: str,
                 self.channel = grpc.insecure_channel(target=self.grpc_address, options=grpc_options,
                                                      compression=grpc_compression)
 
-            # with grpc.secure_channel it takes a lot of time to check the grpc-connection
```
Is this error a reproducible one to leave it here?

Valenzione commented:
> Great work 👍 .


@trthhrtz Would you mind changing your review status from "Request Changes" to "Approved" then?) 🙂

trthhrtz commented:
@Valenzione yeah, I just need to get replies from @KineticCookie and ready to approve

trthhrtz commented:
Did minor refactoring, resolved conflicts @Valenzione

trthhrtz commented:
👍

tidylobster commented on hydrosdk/application.py:
```diff
@@ -22,135 +22,177 @@ class ApplicationStatus(Enum):
+    >>> mv1 = ModelVersion.find(cluster, "my-model", 1)
+    >>> mv2 = ModelVersion.find(cluster, "my-model", 2)
+    >>> stage = ExecutionStageBuilder() \
+                    .with_model_variant(mv1, 50) \
```
Every stage can be comprized of multiple model versions. We call them model variants only within stages.

tidylobster commented on hydrosdk/contract.py:
```diff
@@ -449,10 +386,20 @@ def build(self) -> ModelSignature:
-
-    def __with_field(self, collection, name, dtype, shape, profile=ProfilingType.NONE):
+                    
+    def __with_field(self, collection: List[ModelField], name: str, 
```
Left this as it was before. We have to make a convention regarding single/double underscore notations.
